From ca88b1e42082f48527c8bd7cf6cb2bdc57d6041d Mon Sep 17 00:00:00 2001
From: GUO Zihua <guozihua@huawei.com>
Date: Tue, 19 Jul 2022 22:10:06 +0800
Subject: [PATCH 3607/4585] [Backport] KEYS: Fix error path return value in
 pgp_generate_fingerprint

hulk inclusion
category: bugfix
bugzilla: https://gitee.com/openeuler/kernel/issues/I5GRSV

Reference: https://gitee.com/openeuler/kernel/commit/21f33dad65b601629e5d4b1441aa4b3596c2ef39

--------------------------------

In function pgp_generate_fingerprint, return value is not set correctly
shall kmalloc failed, leading to the caller triggering a read out-of-bound
while reading ctx.fingerprint or ctx.raw_fingerprint.

This patch fixes this issue by setting correct return value on the error
path.

Fixes: 4006f47d4e21 ("KEYS: PGP data parser")
Signed-off-by: GUO Zihua <guozihua@huawei.com>
Reviewed-by: Roberto Sassu <roberto.sassu@huawei.com>
Signed-off-by: Zheng Zengkai <zhengzengkai@huawei.com>
Signed-off-by: GUO Zihua <guozihua@huawei.com>
---
 crypto/asymmetric_keys/pgp_public_key.c | 8 ++++++--
 1 file changed, 6 insertions(+), 2 deletions(-)

diff --git a/crypto/asymmetric_keys/pgp_public_key.c b/crypto/asymmetric_keys/pgp_public_key.c
index 27b9efeafc4f..33a089797e59 100644
--- a/crypto/asymmetric_keys/pgp_public_key.c
+++ b/crypto/asymmetric_keys/pgp_public_key.c
@@ -152,8 +152,10 @@ static int pgp_generate_fingerprint(struct pgp_key_data_parse_context *ctx,
 	digest_size = crypto_shash_digestsize(tfm);
 
 	raw_fingerprint = kmalloc(digest_size, GFP_KERNEL);
-	if (!raw_fingerprint)
+	if (!raw_fingerprint) {
+		ret = -ENOMEM;
 		goto cleanup_hash;
+	}
 
 	ret = crypto_shash_final(digest, raw_fingerprint);
 	if (ret < 0)
@@ -161,8 +163,10 @@ static int pgp_generate_fingerprint(struct pgp_key_data_parse_context *ctx,
 
 	ctx->fingerprint_len = digest_size * 2;
 	fingerprint = kmalloc(digest_size * 2 + 1, GFP_KERNEL);
-	if (!fingerprint)
+	if (!fingerprint) {
+		ret = -ENOMEM;
 		goto cleanup_raw_fingerprint;
+	}
 
 	offset = digest_size - 8;
 	pr_debug("offset %u/%u\n", offset, digest_size);
-- 
2.17.1

